### PR TITLE
Handle nullable annotations on user defined operators and conversions

### LINKED ISSUE #32671: Annotations do not work for user defined conversion operators
**Version Used**: 

Microsoft Visual Studio Professional 2017 
Version 15.2 (26430.16) Release
VisualStudio.15.Release/15.2.0+26430.16
Microsoft .NET Framework
Version 4.7.03056

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
using System.Runtime.CompilerServices;

class C
{
  void M(C? c)
  {
    if (c)
    {
      c.ToString();
    }
  }

  public static extern implicit operator bool([NotNullWhenTrue] C? c);
}

namespace System.Runtime.CompilerServices
{
  class NotNullWhenTrueAttribute : System.Attribute { }
}
```

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWACgABABgqYEYA6AJX1xgBLMHC4BhAPZgADoPIIAyogBuggMZwoAbkaMmAJgpjGAb0YV2AFgoBZABRiA/BTUBKcxTMMLFwQDMKOzcPCy8fHzUuABUJBRgEQVwAcztXHW8fAF8PbIYPJgBmdg4ANioAD3gEXAphaUJ1QRgKCWlECBgJBApgCQlCOwBtADkJGGGiQgB1AAtqKIR8OABdI2c3dNzGGhEoaQgNYt5+IRFxKVl5JQRVDShTfMNR8cnZ+cW4AEEYeMFgfHgFBAR2+v3+gJMFFymSAA==

**Expected Behavior**:
No warnings for `c.ToString()` as according to the annotation it cannot be null when the conversion operator returns `true`

**Actual Behavior**:
`warning CS8602: Possible dereference of a null reference.`

[jcouv update:] We should also test `[NotNullIfNotNull]`, `[DoesNotReturn]`, etc

### LINKED ISSUE #49575: Postconditions on user-defined conversion parameters are not respected
**Version Used**: dc6f316a5e4f8eb3a0d36ab356c5c12686db00a9

**Steps to Reproduce**:

```cs
#nullable enable

using System.Diagnostics.CodeAnalysis;

public static class Program
{
    public static void M1(string? s)
    {
        AssertNotNull(s);
        s.ToString();
    }

    public static void M2(string? s)
    {
        C c = s;
        s.ToString(); // 1
    }

    public static void AssertNotNull([AllowNull, NotNull] string s) { throw null!; }

    public class C
    {
        public static implicit operator C([AllowNull, NotNull] string s) { throw null!; }
    }
}
```

[SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4GIB2BXANliwtwAEcaeBAsAFBUACADITQIwB0AIgJYQDmaA9rA4BjKCwDCfACZwAgqSwBPKBygBuKrQBMAZioBvKoSONtjJgDZGAFkIBZJgApmdAPyEoASkPGDlY/8IZKChEGAA5PnDsLAdPdT8Ao1EAFT4AZRgEDjRuBw94/wBfDQSjGlNmSxobW00nJld3L1LCX0SjMUIhQgBedwL2lPTM7Nz8wgB6CcImbyNi6hbysyqbIJCEcMiw6IcAbRkcPgB3HZwAGkIIqJwAXTMGT1bCGAALBBPCTBwAQlVCBZzEyMTSEMRAtrtZaVQgcMAABywwg4MEIfHhiAgMD4CDB+0OWBOZywl2uxPuziazzeH2OX2ifwBQIWCyAA=)

**Expected Behavior**: No diagnostics

**Actual Behavior**: Diagnostic indicated by `// 1`.

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #34556 (replacing an old PR)
Closes #32671
Closes #49575

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -4385,6 +4390,9 @@ bool learnFromConditionalAccessOrBoolConstant()
                         useLegacyWarnings: false,
                         AssignmentKind.Argument,
                         parameter);
+                    CheckDisallowedNullAssignment(resultType, parameterAnnotations, expr.Syntax.Location, operand);
+
+                    LearnFromPostConditions(operand, GetParameterAnnotations(parameter));
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:4395]:
```suggestion
                    LearnFromPostConditions(operand, parameterAnnotations);
```

**Comment by cston** [comment]
>                     rightState = VisitOptionalImplicitConversion(right, targetTypeOpt: discarded ? default : leftLValueType, UseLegacyWarnings(left, leftLValueType), trackMembers: true, AssignmentKind.Assignment);

Do we need `Unsplit()` after this call?

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:8771 in ecfe673. [](commit_id = ecfe67332f579c70cc8567173c9258be7f66ebd2, deletion_comment = False)

**Comment by cston** [comment]
>                             valueType = VisitOptionalImplicitConversion(rightPart, lvalueType, useLegacyWarnings: true, trackMembers: true, AssignmentKind.Assignment);

`Unsplit()`?

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:9092 in ecfe673. [](commit_id = ecfe67332f579c70cc8567173c9258be7f66ebd2, deletion_comment = False)

**Review by cston — APPROVED** [review]

**Comment by RikkiGibson** [comment]
I couldn't think of ways that the conditional state from the VisitOptionalImplicitConversion would be meaningfully used at either of those locations, so I added the Unsplit() call to be "conservative" about it.

**Review by chsienki — APPROVED** [review]

**Pull request merged**